re-enable proxy service fails on execstart integration test

engelmi opened the pull request:
Fixes: #320 

This PR re-enables the - currently - skipped integration test for the proxy service where the `ExecStart` returns exit code 1. 
Running it locally a few times it works fine, so lets enable it again and see if the flakiness is gone.

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -136,13 +136,15 @@ def wait_for_unit_state_to_be(
+            state = self.get_unit_state(unit_name)
+            LOGGER.info(f"Found latest state: {state}")
+            if state == expected_state:
                 return True
```
Is there a reason to loose debug logging by not using private function `_is_unit_in_state()`? I've extracted the comparison along with logging to prevent code duplication and having detailed logging to to check status change progress

engelmi replied:
Locally it was running fine, so I assumed the flakyness has been fixed by chance. This PR should still be in draft mode, I'll have to debug this - the log change was just for that.

Also, we log on info level, so the debug statements arent shown. Could we change this to use debug by default?

mwperina replied:
You can change the default log level:
https://github.com/eclipse-bluechi/bluechi/tree/main/tests#changing-log-level

I intentionally changed the most important log messages to INFO, because on DEBUG there are quite a lot debug messages from podman-py, which makes the log hard to read. If you think the level in `_is_unit_in_state` should be on INFO, feel free to change it

engelmi replied:
Changed it to info level. Otherwise, if an assertion on the expected state times out, there is no information what the last state was.

mwperina commented on tests/tests/tier0/proxy-service-fails-on-execstart/test_proxy_service_fails_on_execstart.py:
```diff
@@ -24,7 +23,7 @@ def verify_proxy_start_failed(foo: BluechiNodeContainer, bar: BluechiNodeContain
     assert bar.wait_for_unit_state_to_be(simple_service, "failed")
     bluechi_dep_service = assemble_bluechi_dep_service_name(simple_service)
-    assert bar.wait_for_unit_state_to_be(bluechi_dep_service, "inactive")
+    assert bar.wait_for_unit_state_to_be(bluechi_dep_service, "failed")
```
According to the logs the state reported was `inactive`:

https://github.com/eclipse-bluechi/bluechi/actions/runs/6717283203/job/18255091471?pr=627#step:9:6934

Above you have changed the code from `inactive` to `failed`, so which status do we want to wait for?

engelmi replied:
The state of the bluechi-dep@.sevice is the cause for the flakyness... which I cant repdroduce locally. Its either in inactive or failed state, I'll need to debug this in the CI unfortunately. ~It should be in a failed state.~

engelmi replied:
The `bluechi-dep@simple.service` should be in state `inactive`. For some reason, running locally
- the same test ~10 times 
- using the same setup (2 container, same systemd units) and starting `requesting.service` 20 times

I wasn't able to get it into a `failed` state. In the CI its often failed, sometimes inactive. 

Looking at the logs for the node bar 
- when it is in `failed` state:
```
Nov 01 07:08:34 27662043abed bluechi-agent[27]: Stopping dependency bluechi-dep@simple.service
Nov 01 07:08:34 27662043abed systemd[1]: bluechi-dep@simple.service: Main process exited, code=killed, status=15/TERM
Nov 01 07:08:34 27662043abed systemd[1]: bluechi-dep@simple.service: Failed with result 'signal'.
Nov 01 07:08:34 27662043abed systemd[1]: Stopped BlueChi proxy dependency service.
```
- when it is in `inactive` state:
```
Nov 01 08:02:25 0e0f91f7e4e4 bluechi-agent[27]: Sending UnitPropertiesChanged simple.service
Nov 01 08:02:25 0e0f91f7e4e4 systemd[1]: bluechi-dep@simple.service: Deactivated successfully.
Nov 01 08:02:25 0e0f91f7e4e4 systemd[1]: Stopped BlueChi proxy dependency service.
Nov 01 08:02:25 0e0f91f7e4e4 bluechi-agent[27]: Stopping dependency bluechi-dep@simple.service
```

I noticed that for the expected `inactive` state, the call from BlueChi to stop the `bluechi-dep@.service` comes in when the service has already been stopped (due to the `BindsTo=`). And in the `failed` case it arrives before - which leads to the `SIGTERM`, I assume. So I further assume that this is some kind of race condition with BlueChi and systemd stopping the service. Reading through the documentation: 
https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecStop=
I think that adding a simple `ExecStop=true` will prevent systemd from sending `SIGTERM` to the dep service - thus preventing the `failed` state. 

Could you have a look and confirm/refute? @alexlarsson

engelmi committed:
re-enable proxy service fails on execstart integration test

Signed-off-by: Michael Engel <mengel@redhat.com>

engelmi committed:
replaced ExecStart= with ExecStop= in bluechi-dep@.service

By using ExecStop= instead of ExecStart= a race condition is
avoided that leads systemd to send a SIGTERM signal to the
dep service, resulting in it being in a failed state.

Signed-off-by: Michael Engel <mengel@redhat.com>

alexlarsson commented on systemd-units/bluechi-dep@.service:
```diff
@@ -10,5 +10,6 @@ After=%i.service
 
 [Service]
 ExecStart=true
+ExecStop=true
```
I don't think this actually fully solves the race condition. systemd will first run ExecStop commands, but unless the process died it will still send kill signals after that.

alexlarsson replied:
Maybe `KillMode=none` instead?

engelmi replied:
Yes, it didn't fully solve the race condition. 
I'd use your proposed solution to just use `ExecStop=/bin/true`. I think additionally having `SuccessExitStatus=` doesn't hurt, but also shouldn't have any effect - so just sticking with `ExecStop=` should be best. What do you think? @alexlarsson

alexlarsson commented on systemd-units/bluechi-dep@.service:
```diff
@@ -9,6 +9,6 @@ BindsTo=%i.service
 [Service]
-ExecStart=true
 Type=oneshot
+ExecStop=true
```
Maybe a comment about why it is done this way?

engelmi replied:
Good point! Added :heavy_check_mark:

mwperina reviewed (approved):
LGTM
